arbnode/maintenance.go

```diff
@@ -120,157 +72,92 @@ func NewMaintenanceRunner(config MaintenanceConfigFetcher, seqCoordinator *SeqCo
-		return true, time.Time{}
-	}
-	return false, time.UnixMilli(milli)
-}
-
-// bool if running currently, if false - duration since last time it was running
-func (mr *MaintenanceRunner) TimeSinceLastMaintenance() (bool, time.Duration) {
-	running, maintTime := mr.getPrevMaintenance()
-	if running {
-		return true, 0
-	}
-	return false, time.Since(maintTime)
-}
-
-func (mr *MaintenanceRunner) setMaintenanceDone() {
-	milli := time.Now().UnixMilli()
-	prev := mr.lastMaintenance.Swap(milli)
-	if prev != 0 {
-		log.Error("maintenance executed in parallel", "current", time.UnixMilli(milli), "prev", time.UnixMilli(prev))
-	}
-}
-
-func (mr *MaintenanceRunner) setMaintenanceStart() error {
-	prev := mr.lastMaintenance.Swap(0)
-	if prev == 0 {
-		return errors.New("already running")
-	}
-	return nil
-}
-
-func (mr *MaintenanceRunner) maybeRunScheduledMaintenance(ctx context.Context) time.Duration {
+// exported for testing
+func (mr *MaintenanceRunner) MaybeRunMaintenance(ctx context.Context) time.Duration {
 	config := mr.config()
-	if !config.enabled {
-		return time.Minute
+	if !config.Enable {
+		log.Debug("maintenance is disabled, skipping")
+		return config.CheckInterval
 	}
 
-	now := time.Now().UTC()
-
-	inMaintenance, lastMaintenance := mr.getPrevMaintenance()
-	if inMaintenance {
-		return time.Minute
-	}
-
-	if !wentPastTimeOfDay(lastMaintenance, now, config.minutesAfterMidnight) {
-		return time.Minute
-	}
-
-	err := mr.attemptMaintenance(ctx)
+	shouldTriggerMaintenance, err := mr.exec.ShouldTriggerMaintenance().Await(mr.GetContext())
 	if err != nil {
-		log.Warn("scheduled maintenance error", "err", err)
+		log.Error("error checking if maintenance should be triggered", "err", err)
+		return config.CheckInterval
 	}
-
-	return time.Minute
-}
-
-func (mr *MaintenanceRunner) Trigger() error {
-	if !mr.config().Triggerable {
-		return errors.New("maintenance not configured to be triggerable")
+	if !shouldTriggerMaintenance {
+		log.Debug("skipping maintenance, not triggered")
+		return config.CheckInterval
 	}
-	if running, _ := mr.getPrevMaintenance(); running {
-		return nil
-	}
-	// maintenance takes a long time, run on a separate thread
-	mr.LaunchThread(func(ctx context.Context) {
-		err := mr.attemptMaintenance(ctx)
-		if err != nil {
-			log.Warn("triggered maintenance returned error", "err", err)
-		}
-	})
-	return nil
-}
 
-func (mr *MaintenanceRunner) attemptMaintenance(ctx context.Context) error {
+	// If seqCoordinator is nil there is no need to coordinate maintenance running with other sequecers.
 	if mr.seqCoordinator == nil {
-		return mr.runMaintenance()
+		mr.runMaintenance()
+		return config.CheckInterval
 	}
 
 	release := make(chan struct{})
 	if !mr.lock.AttemptLockAndPeriodicallyRefreshIt(ctx, release) {
-		return errors.New("did not catch maintenance lock")
+		log.Warn("maintenance lock not acquired, skipping maintenance")
+		return config.CheckInterval
 	}
 	defer func() {
-		release <- struct{}{}
+		select {
```

Comment:
If I understand correctly and that's the deadlock - I think a close(release) will work better.

Reply:
Done.